scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -1667,3 +1670,459 @@ cdef FibonacciNode* remove_min(FibonacciHeap* heap) noexcept:
+                for i in range(candidate_distances.shape[0]):
+                    if candidate_distances[i] == INFINITY:
+                        short_path_idx = i
+                        break
+                    elif candidate_distances[i] > tmp_d:
+                        tmp_d = candidate_distances[i]
+                        tmp_i = i
+                if short_path_idx ==  NULL_IDX and total_distance < tmp_d:
+                    short_path_idx = tmp_i
+
+                if short_path_idx != NULL_IDX:
+                    candidate_distances[short_path_idx] = total_distance
+                    # Reset candidate_predecessors[short_path_idx]
+                    candidate_predecessors[short_path_idx, :] = NULL_IDX
+                    # Fill original path
+                    node = spur_node
+                    while node != NULL_IDX:
+                        candidate_predecessors[short_path_idx, node] = (
+                            shortest_paths_predecessors[k-1, node]
+                        )
+                        node = shortest_paths_predecessors[k-1, node]
+
+                    # Fill spur path
+                    node = sink
+                    while node != spur_node:
+                        candidate_predecessors[short_path_idx, node] = (
+                            predecessor_matrix[0, node]
+                        )
+                        node = predecessor_matrix[0, node]
+
+           # ---------------------------------------------------
+            # Restore graph weights
+            node = spur_node
+            while node != NULL_IDX:
+                csr_weights[csr_indptr[node]: csr_indptr[node + 1]] = (
+                    original_weights[csr_indptr[node]: csr_indptr[node + 1]]
+                )
+                if not directed:
+                    csrT_weights[csrT_indptr[node]: csrT_indptr[node + 1]] = (
+                        originalT_weights[csrT_indptr[node]: csrT_indptr[node + 1]]
+                    )
+                node = shortest_paths_predecessors[k-1, node]
+
+
+        # ---------------------------------------------------
+        # Find shortest path in candidates and add to result arrays
+        tmp_d = INFINITY # Minimal distance in potential distances array
+        short_path_idx = NULL_IDX
+        for i in range(candidate_distances.shape[0]):
+            if candidate_distances[i] < tmp_d:
+                tmp_d = candidate_distances[i]
+                short_path_idx = i
+        if short_path_idx == NULL_IDX:
+            # There are no more paths
+            break
+        else:
+            shortest_distances[k] = candidate_distances[short_path_idx]
+            # Remove path from candidates and add to shortest_paths_predecessors
+            candidate_distances[short_path_idx] = INFINITY
+            shortest_paths_predecessors[k] = candidate_predecessors[short_path_idx]
+
+
+@cython.boundscheck(False)
+cdef bint _yen_is_path_in_candidates(
```

Comment:
We need to store edges to avoid searching into previously computed paths to know which edges were removed before. 

Suppose you want to compute the deviations of a path `P` and you know that `P` has itself been obtained as a deviation from spur node `u`. You don't know which path `Q` was considered at the time and so which edge was removed. Furthermore, `Q` might be the result of a deviation from another path at spur node `u` and so again removing some edge, etc.  To compute the deviations of `P` at spur node `u`, we must remove all these edges since otherwise we may (will) find a previously computed path.

Said differently, Yen's algorithm scans the list of previously computed path with same prefix to identify edges that should not be considered when computing the deviations of `P` at `u`. Storing these edges avoids to scan these paths. 

And for sure, we don't want to find twice a path.

Reply:
Thank you for the elaborate answer.

For every spur node, we scan the list of previously found shortest paths. In case the "original path" (i.e., the path up to the spur node) coincides with a known short path, we remove the next edge in that path.
Sorry, but I still don't understand why this is not sufficient. Or why your suggestion is different from the original algorithm in more than some spur nodes being filtered out since we know they will not yield the shortest path (which is a great optimization by itself!)?

Of course, later we also remove the rest of the original path up to the spur node. But that doesn't matter to the discussion (I think).

Reply:
> Thank you for the elaborate answer.
> 
> For every spur node, we scan the list of previously found shortest paths. In case the "original path" (i.e., the path up to the spur node) coincides with a known short path, we remove the next edge in that path. Sorry, but I still don't understand why this is not sufficient. Or why your suggestion is different from the original algorithm in more than some spur nodes being filtered out since we know they will not yield the shortest path (which is a great optimization by itself!)?

During the execution of the algorithm, you may first find a candidate path P1 with prefix Q (that is a path from the source to a node u different from the sink). latter you find another path P2 with same prefix Q. 
Next time you will have to compute a deviation for a path with prefix Q from spur node u (the last vertex of u), you have to remove both the next edge in P1 and the next edge in P2.
To find these edges, either you search for all previously computed paths to identify those with prefix Q and the edges you need to remove. The alternative, is to store for each candidate path: the path itself + the spur node when this path has been computed + the edges that were removed. This way, when you extract a candidate path as ith shortest path, you have all you need to compute its deviations.

This will not reduce the computational complexity of the algorithm, but save time in practice.
However, I agree that it is slightly more complex to implement.

> Of course, later we also remove the rest of the original path up to the spur node. But that doesn't matter to the discussion (I think).

Indeed.